### PR TITLE
[DAGCombine] Count leading ones: refine post DAG/Type Legalisation if promotion

### PR DESCRIPTION
**Pull request by v01dXYZ**
This PR is related to #99591. In this PR, instead of modifying how the legalisation occurs depending on surrounding instructions, we refine after legalisation.

This PR has two parts:

* `SDPatternMatch/MatchContext`: Modify a little bit the code to match Operands (used by `m_Node(...)`) and Unary/Binary/Ternary Patterns to make it compatible with `VPMatchContext`, instead of only `m_Opc` supported. Some tests were added to ensure no regressions. 
* `DAGCombiner`: Add a `foldSubCtlzNot` which detect and rewrite the patterns using matching context.

Remaining Tasks:

- [ ] GlobalISel
- [ ] Currently the pattern matching will occur even before legalisation. Should I restrict it to specific stages instead ?
- [ ] Style: Add a visitVP_SUB ?? Move `foldSubCtlzNot` in another location for style consistency purpose ?

@topperc

### CODE REVIEW

**Comment by mshockwave** [comment]
could you split this PR into at least two separate ones

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 1]
```diff
@@ -464,7 +467,7 @@ struct TernaryOpc_match {
   bool match(const MatchContext &Ctx, SDValue N) {
     if (sd_context_match(N, Ctx, m_Opc(Opcode))) {
       EffectiveOperands<ExcludeChain> EO(N);
-      assert(EO.Size == 3);
+      assert(EO.Size == 3U + 2 * N->isVPOpcode());
```
Comment (mshockwave) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:470]: would it be better if we put this into `EffectiveOperands`? `EffectiveOperands` was invented to abstract away "don't care" operands, like input chain, under certain circumstances so I feel like it fits well with what you're doing here.

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 2]
```diff
@@ -390,7 +392,8 @@ template <unsigned OpIdx, typename... OpndPreds> struct Operands_match {
   template <typename MatchContext>
   bool match(const MatchContext &Ctx, SDValue N) {
     // Returns false if there are more operands than predicates;
-    return N->getNumOperands() == OpIdx;
+    // Ignores the last two operands if both the Context and the Node are VP
+    return N->getNumOperands() == (OpIdx + 2 * Ctx.IsVP * N->isVPOpcode());
```
Comment (mshockwave) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:396]: readability: maybe something like `Ctx.IsVP && N->isVPOpcode() ? OpIdx + 2 : OpIdx`

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 3]
```diff
@@ -516,7 +519,7 @@ struct BinaryOpc_match {
   bool match(const MatchContext &Ctx, SDValue N) {
     if (sd_context_match(N, Ctx, m_Opc(Opcode))) {
       EffectiveOperands<ExcludeChain> EO(N);
-      assert(EO.Size == 2);
+      assert(EO.Size == 2U + 2 * N->isVPOpcode());
```
Comment (mshockwave) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:522]: ditto

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 4]
```diff
@@ -668,7 +671,7 @@ template <typename Opnd_P, bool ExcludeChain = false> struct UnaryOpc_match {
   bool match(const MatchContext &Ctx, SDValue N) {
     if (sd_context_match(N, Ctx, m_Opc(Opcode))) {
       EffectiveOperands<ExcludeChain> EO(N);
-      assert(EO.Size == 1);
+      assert(EO.Size == 1U + 2 * N->isVPOpcode());
```
Comment (mshockwave) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:674]: ditto

File: llvm/include/llvm/CodeGen/SDPatternMatch.h [hunk 5]
```diff
@@ -47,6 +47,8 @@ class BasicMatchContext {
   bool match(SDValue N, unsigned Opcode) const {
     return N->getOpcode() == Opcode;
   }
+
+  static constexpr bool IsVP = false;
```
Comment (mshockwave) [inline llvm/include/llvm/CodeGen/SDPatternMatch.h:51]: have you explored the ideas of abstracting the number of operands instead? Specifically, having a match context to implement its own version of `unsigned getNumOperands(SDValue N)` (or something similar)?

**Comment by v01dXYZ** [comment]
> could you split this PR into at least two separate ones

I'll do that as the SDPatternMatch seems to require work on its own.

**Comment by v01dXYZ** [comment]
Putting this PR back to Draft as it is blocked by the PR for SDPatternMatch.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 6]
```diff
@@ -3755,6 +3755,51 @@ SDValue DAGCombiner::foldSubToUSubSat(EVT DstVT, SDNode *N, const SDLoc &DL) {
...
+  APInt AndMask;
+  APInt XorMask;
+  APInt BitWidthDiff;
+
+  SDValue CtlzOp;
+  SDValue Src;
+
+  if (!sd_context_match(
+          N, Matcher,
+          m_Sub(m_Node(ISD::CTLZ, m_Value(CtlzOp)), m_ConstInt(BitWidthDiff))))
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:3776]: I'd expect to have an m_Ctlz wrapper for m_Node(ISD::CTLZ

Comment (v01dXYZ) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:3776]: The commit is not complete. I am implenting tests

**Review by arsenm — APPROVED** [review]

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 7]
```diff
@@ -3755,6 +3755,50 @@ SDValue DAGCombiner::foldSubToUSubSat(EVT DstVT, SDNode *N, const SDLoc &DL) {
   return SDValue();
 }
 
+template <class MatchContextClass>
+static SDValue foldSubCtlzNot(SDNode *N, SelectionDAG &DAG) {
```
Comment (mshockwave) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:3759]: could you add a brief comment explaining what this function does, primarily the pattern?

**Review by mshockwave — APPROVED** [review]
LGTM w/ minor comment

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 7 above)
Comment (v01dXYZ) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:3759]: I added a comment. There are two patterns that are matched. And the replacing subDAG is slightly different (one any-extend when the other one does not).

Comment (v01dXYZ) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:3759]: Ping

**Review by arsenm — APPROVED** [review]

**Pull request merged**